tests/test_directives.py

```diff
@@ -0,0 +1,180 @@
+    #     for directive in available_directives_list:
+    #         directive.validation(directive.default)
+
+    def test_validation_manual(self):
+        directive_copy = deepcopy(NP)
+
+        def div_by_ten(v):
+            if v % 10 == 0:
+                return True
+            return False
+
+        directive_copy.validation = div_by_ten
+        assert directive_copy.validation(10)
+        assert not directive_copy.validation(5)
+        with pytest.raises(TypeError):
+            directive_copy.validation('4')
+
+    def test_serial(self):
+        assert NP.serial(4, 2) == 4
+        assert NRANKS.serial(4, 2) == 4
+        assert NGPU.serial(4, 2) == 4
+        assert OMP_NUM_THREADS(4, 2) == 4
+        assert EXECUTABLE.serial('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.serial(4, 2) == 6
+        assert MEMORY.serial(4, 2) == 4
+        assert PROCESS_FRACTION(4.0, 2.0) == 4.0
+
+    def test_parallel(self):
+        assert NP.parallel(4, 2) == 6
+        assert NRANKS.parallel(4, 2) == 6
+        assert NGPU.parallel(4, 2) == 6
+        assert OMP_NUM_THREADS(4, 2) == 6
+        assert EXECUTABLE.parallel('Path1', 'Path2') == 'Path1'
+        assert WALLTIME.parallel(4, 2) == 4
+        assert MEMORY.parallel(4, 2) == 6
+        assert PROCESS_FRACTION.parallel(4.0, 2.0) == 4.0
+
+    def test_parallel_manual(self):
+        directive_copy = deepcopy(NP)
+        directive_copy.parallel = lambda x, y: x*y
+        parallel = directive_copy.parallel(3, 2)
+        assert parallel == 6
+        directive_copy.validation(parallel)
+        with pytest.raises(TypeError):
+            directive_copy.parallel('4', '2')
+
+    def test_finalize(self):
+        dict_directives = {'nranks': NRANKS.default, 'omp_num_threads': OMP_NUM_THREADS.default}
+        assert NP.finalize(2, dict_directives) == 2
+        dict_directives['nranks'] = 2
+        dict_directives['omp_num_threads'] = 4
+        assert NP.finalize(2, dict_directives) == 8
+        dict_directives['nranks'] = NRANKS
+        assert NP.finalize(2, dict_directives) == 2
+
+    def test_manual_item(self, setUp):
+        def val(v):
+            if v < 10:
+                raise ValueError("Price cannot be less than 10 units")
+            return v
+
+        def no_aggregation(v, o):
+            return v
+
+        def finalize(value, dict):
+            discount = dict.get('discount', 0)
+            free = dict.get('free', False)
+            value = value - discount
+            if value < 0 or free:
+                return 0
+            return value
+
+        PRODUCT = setUp(_DirectivesItem, name='product', validation=val,
+                        default=10, serial=no_aggregation, parallel=no_aggregation,
+                        finalize=finalize)
+        asset_dict = {'free': False, 'discount': 5}
+        assert PRODUCT.default == PRODUCT.validation(PRODUCT.default)
+        assert PRODUCT.default == PRODUCT.validation(20)
```

Comment:
This should not be the case right. `PRODUCT.default` should be 10 right?

Reply:
Yes, this was a typographical error. I'll correct this.